### ISSUE TITLE
train yolo with random weighted sampler

### ISSUE THREAD

**Issue by Dreahim** [ultralytics/ultralytics#18885]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have many data sources, Now I can write all data paths in data.yaml file and when I start the training, the different dataset sources are merged together into only 1 dataset.
Can I use or create a method like random weighted sampler to take equal number of samples from each data source for every epoch

For example:
data1 (20,000 images)
data2 (1000 images)

I want every epoch to have only 1000 images from each data source instead of having all images for every epoch

How can I do this or what are the methods I should edit to do this

thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Dreahim, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples, training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For your specific use case, it sounds like you're looking to implement a custom sampling method to balance data sources during training. This is not directly supported in the current version. However, it is feasible to modify the dataset loading behavior or create a custom dataloader to incorporate a random weighted sampler. If you'd like guidance or suggestions on implementing this, feel free to share more details, and our team will assist soon!

### Community Resources

In the meantime, you might find ideas or inspiration from others in the community to solve this. Join us here:
- 🎧 [Discord](https://discord.com/invite/ultralytics) for real-time chat.
- 🗨️ [Discourse](https://community.ultralytics.com/) for more in-depth discussions.
- 🕵️‍♂️ [Subreddit](https://reddit.com/r/Ultralytics) to share and explore solutions.

## Upgrade

Make sure you're running the latest version of the `ultralytics` package for the best experience. Upgrade it along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment as follows:

```bash
pip install -U ultralytics
```

## Environments

Here are some pre-configured, efficient environments to run YOLO training and experiments (with all dependencies preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

⚡ This is an automated response, but don't worry, an Ultralytics engineer will respond to you directly as soon as possible! 😊

**Comment by Y-T-G** [comment]
You can modify build.py with your dataloader


https://github.com/ultralytics/ultralytics/blob/de05d1b65509e14ff9396a3aa5ba8a958bcedcf9/ultralytics/data/build.py

**Comment by yeonhyochoi** [comment]
> You can modify build.py with your dataloader
> 
> https://github.com/ultralytics/ultralytics/blob/de05d1b65509e14ff9396a3aa5ba8a958bcedcf9/ultralytics/data/build.py

hello. i want to try similiar task.

#18809 

can you give me advice? 
i think the sampler is not updated properly, but, i couldnt fix it.

**Comment by glenn-jocher** [comment]
@yeonhyochoi thank you for your question. If you're encountering issues with the sampler not updating properly, I recommend reviewing the `_RepeatSampler` class in `ultralytics/data/build.py` since it manages the iteration logic. Ensure your custom sampler logic integrates correctly with the `InfiniteDataLoader`. For further debugging, you may also log and verify the behavior of the `sampler` and `batch_sampler` during training to pinpoint the issue. If you have specific code snippets or error logs, feel free to share them for more targeted guidance.

**Comment by yeonhyochoi** [comment]
i tried like this, 
when epoch updated, new loader load.
but it trained very slowly and, stop at second epoch without error log.
after 1 hours, the train process was proceeded, but very very slow..
i think, it is not good idea upated train_loader in train loop. 
![Image](https://github.com/user-attachments/assets/8f36fa62-cbbc-407f-bcac-6c8569d4e8a0)

**Comment by yeonhyochoi** [comment]
i tried like this,
when epoch updated, new loader load.
but it trained very slowly and, stop at second epoch without error log.
after 1 hours, the train process was proceeded, but very very slow..
i think, it is not good idea upated train_loader in train loop.

![Image](https://github.com/user-attachments/assets/0d5f2b41-e2e4-4445-bd88-57c1a9852a05)

**Comment by yeonhyochoi** [comment]
> [@yeonhyochoi](https://github.com/yeonhyochoi) thank you for your question. If you're encountering issues with the sampler not updating properly, I recommend reviewing the `_RepeatSampler` class in `ultralytics/data/build.py` since it manages the iteration logic. Ensure your custom sampler logic integrates correctly with the `InfiniteDataLoader`. For further debugging, you may also log and verify the behavior of the `sampler` and `batch_sampler` during training to pinpoint the issue. If you have specific code snippets or error logs, feel free to share them for more targeted guidance.

can you check https://github.com/ultralytics/ultralytics/issues/18809 ? i posted my attempt trial

**Comment by glenn-jocher** [comment]
@yeonhyochoi thank you for sharing your update. I will review the details in [#18809](https://github.com/ultralytics/ultralytics/issues/18809) to better understand your approach and provide specific feedback. Please ensure to include any relevant code snippets or logs there to help diagnose the issue effectively.
